File: apex_dashboard/apex_dashboard/page/employee_custody_dashboard/employee_custody_dashboard.py
from __future__ import annotations

from typing import Dict, List, Optional

import frappe
from frappe import _
from frappe.utils import getdate, nowdate

from apex_dashboard.dashboard import utils as dashboard_utils


@frappe.whitelist()
def get_context(context):
	context.no_cache = 1
	context.page_title = _("Employee Custody Dashboard")
	return context


@frappe.whitelist()
def get_dashboard_data(company: Optional[str] = None, posting_date: Optional[str] = None):
	try:
		posting_date = posting_date or nowdate()
		posting_date = str(getdate(posting_date))

		data = build_employee_custody_snapshot(company=company, posting_date=posting_date)

		return {
			"success": True,
			"filters": {
				"company": company,
				"posting_date": posting_date,
			},
			"data": data,
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "Employee Custody Dashboard - get_dashboard_data")
		return {
			"success": False,
			"error": _(str(exc)),
		}


def build_employee_custody_snapshot(company: Optional[str], posting_date: str) -> Dict[str, object]:
	mapping = dashboard_utils.get_accounts_for_section("employee_custody_dashboard")

	custody = dashboard_utils.get_balances_for_accounts(mapping.get("custody", []), company, posting_date)
	refundable = dashboard_utils.get_balances_for_accounts(mapping.get("refundable_withdrawals", []), company, posting_date)
	cash_advances = dashboard_utils.get_balances_for_accounts(mapping.get("cash_advances", []), company, posting_date)
	consignment = dashboard_utils.get_balances_for_accounts(mapping.get("consignment_goods", []), company, posting_date)
	payables = dashboard_utils.get_balances_for_accounts(mapping.get("payables", []), company, posting_date)

	kpis = [
		{
			"key": "custody",
			"label": _("عهد الموظفين"),
			"totals": _totals(custody),
			"indicator": "warning" if _totals(custody)["base"] > 0 else "info",
		},
		{
			"key": "refundable",
			"label": _("سلف مستردة"),
			"totals": _totals(refundable),
			"indicator": "info",
		},
		{
			"key": "cash_advances",
			"label": _("سلف نقدية"),
			"totals": _totals(cash_advances),
			"indicator": "danger" if _totals(cash_advances)["base"] > 0 else "info",
		},
		{
			"key": "consignment",
			"label": _("بضائع بالعارية"),
			"totals": _totals(consignment),
			"indicator": "warning" if _totals(consignment)["base"] > 0 else "info",
		},
		{
			"key": "payables",
			"label": _("مستحقات الموظفين"),
			"totals": _totals(payables),
			"indicator": "danger" if _totals(payables)["base"] > 0 else "info",
		},
	]

	alerts = _build_alerts(custody, cash_advances, payables)

	return {
		"kpis": kpis,
		"sections": [
			_format_section(_("عهد الموظفين"), custody),
			_format_section(_("سلف مستردة"), refundable),
			_format_section(_("سلف نقدية"), cash_advances),
			_format_section(_("بضائع بالعارية"), consignment),
			_format_section(_("مستحقات الموظفين (خصوم)"), payables),
		],
		"alerts": alerts,
	}


def _totals(balances: List[dashboard_utils.AccountBalance]) -> Dict[str, object]:
	return {
		"by_currency": dashboard_utils.summarize_balances_by_currency(balances),
		"base": dashboard_utils.summarize_base_total(balances),
	}


def _format_section(label: str, balances: List[dashboard_utils.AccountBalance]) -> Dict[str, object]:
	return {
		"label": label,
		"totals": _totals(balances),
		"balances": [balance.__dict__ for balance in balances],
	}


def _build_alerts(
	custody: List[dashboard_utils.AccountBalance],
	cash_advances: List[dashboard_utils.AccountBalance],
	payables: List[dashboard_utils.AccountBalance],
) -> List[Dict[str, object]]:
	alerts: List[Dict[str, object]] = []

	custody_total = _totals(custody)["base"]
	if custody_total > 0:
		alerts.append(
			{
				"level": "warning",
				"message": _("هناك عهد قائمة لم تُسوى بعد، يُنصح بمتابعة الموظفين المعنيين."),
			}
		)

	cash_total = _totals(cash_advances)["base"]
	if cash_total > 0:
		alerts.append(
			{
				"level": "danger",
				"message": _("سلف نقدية مفتوحة تحتاج لتسوية أو خصم قريباً."),
			}
		)

	payable_total = _totals(payables)["base"]
	if payable_total > 0:
		alerts.append(
			{
				"level": "info",
				"message": _("هناك مستحقات رواتب أو عمولات معلقة على الموظفين، تأكد من السداد في الجدول الزمني."),
			}
		)

	if not alerts:
		alerts.append(
			{
				"level": "info",
				"message": _("لا توجد تنبيهات حالية على عهد الموظفين."),
			}
		)

	return alerts

File: apex_dashboard/apex_dashboard/page/employee_custody_dashboard/employee_custody_dashboard.js
frappe.pages["employee-custody-dashboard"].on_page_load = function (wrapper) {
	frappe.require(
		[
			"assets/apex_dashboard/js/dashboard_common.js",
			"assets/apex_dashboard/css/dashboard_common.css",
		],
		() => init_employee_custody_page(wrapper)
	);
};

function init_employee_custody_page(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Employee Custody Dashboard"),
		single_column: true,
	});

	const state = {
		filters: {
			company: frappe.defaults.get_user_default("company"),
			posting_date: frappe.datetime.get_today(),
		},
		data: null,
	};

	const layout = build_employee_custody_layout(page);
	setup_employee_custody_filters(page, state, refresh);

	function refresh() {
		page.set_indicator(__("جارِ التحميل..."), "blue");
		frappe
			.call("apex_dashboard.apex_dashboard.page.employee_custody_dashboard.employee_custody_dashboard.get_dashboard_data", {
				company: state.filters.company,
				posting_date: state.filters.posting_date,
			})
			.then((response) => {
				const payload = response.message || {};
				if (!payload.success) {
					throw new Error(payload.error || __("فشل تحميل بيانات العهد والسلف."));
				}
				state.data = payload.data || {};
				state.filters = Object.assign({}, state.filters, payload.filters || {});
				render_employee_custody_dashboard(layout, state.data);
				page.set_indicator(__("جاهز"), "green");
			})
			.catch((error) => {
				console.error(error);
				page.set_indicator(__("خطأ"), "red");
				frappe.msgprint({
					title: __("خطأ"),
					message: error.message || __("تعذر تحميل لوحة العهد والسلف."),
					indicator: "red",
				});
			});
	}

	refresh();
}

function build_employee_custody_layout(page) {
	const $container = $(page.body);
	$container.empty();

	const $wrapper = $(`
		<div class="gt-dashboard-wrapper">
			<section class="gt-dashboard-section">
				<div class="gt-dashboard-cards" data-role="kpis"></div>
			</section>
			<section class="gt-dashboard-section gt-dashboard-section--split">
				<div class="gt-dashboard-alerts" data-role="alerts"></div>
				<div class="gt-dashboard-section" data-role="summary"></div>
			</section>
			<section class="gt-dashboard-section" data-role="sections"></section>
		</div>
	`);

	$container.append($wrapper);

	return {
		kpis: $wrapper.find('[data-role="kpis"]')[0],
		alerts: $wrapper.find('[data-role="alerts"]')[0],
		sections: $wrapper.find('[data-role="sections"]')[0],
	};
}

function setup_employee_custody_filters(page, state, refresh) {
	const company_field = page.add_field({
		fieldtype: "Link",
		fieldname: "company",
		label: __("شركة"),
		options: "Company",
		default: state.filters.company,
		change() {
			state.filters.company = company_field.get_value();
			refresh();
		},
	});

	const date_field = page.add_field({
		fieldtype: "Date",
		fieldname: "posting_date",
		label: __("تاريخ"),
		default: state.filters.posting_date,
		change() {
			state.filters.posting_date = date_field.get_value();
			refresh();
		},
	});

	page.add_action_icon("refresh", () => refresh());
}

function render_employee_custody_dashboard(layout, data) {
	render_employee_custody_kpis(layout.kpis, data.kpis || []);
	render_employee_custody_alerts(layout.alerts, data.alerts || []);
	render_employee_custody_sections(layout.sections, data.sections || []);
}

function render_employee_custody_kpis(container, kpis) {
	if (!container) return;
	const cards = kpis.map((kpi) => {
		const totals = kpi.totals || {};
		const baseValue = totals.base || 0;
		const byCurrency = totals.by_currency || {};
		const subtitle = Object.keys(byCurrency)
			.map((currency) => DashboardCommon.formatCurrency(byCurrency[currency], currency))
			.join(" • ");

		return {
			title: kpi.label || "",
			value: DashboardCommon.formatCurrency(
				baseValue,
				frappe.defaults.get_default("currency") || frappe.sys_defaults.currency || "EGP"
			),
			subtitle,
			indicator: kpi.indicator || "info",
		};
	});

	DashboardCommon.renderCards(container, cards);
}

function render_employee_custody_alerts(container, alerts) {
	if (!container) return;
	container.innerHTML = "";

	if (!alerts.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد تنبيهات")}</div>`;
		return;
	}

	alerts.forEach((alert) => {
		const item = document.createElement("div");
		item.className = `gt-dashboard-alert level-${alert.level || "info"}`;
		item.innerHTML = `
			<span class="gt-dashboard-alert__badge">${(alert.level || "").toUpperCase()}</span>
			<span class="gt-dashboard-alert__message">${frappe.utils.escape_html(alert.message || "")}</span>
		`;
		container.appendChild(item);
	});
}

function render_employee_custody_sections(container, sections) {
	if (!container) return;
	container.innerHTML = "";

	if (!sections.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات تفصيلية")}</div>`;
		return;
	}

	sections.forEach((section) => {
		const card = document.createElement("div");
		card.className = "gt-dashboard-subcard";

		const header = document.createElement("div");
		header.className = "gt-dashboard-section__header";
		const title = document.createElement("h4");
		title.className = "gt-dashboard-section__title";
		title.innerText = section.label || "";
		header.appendChild(title);
		header.appendChild(DashboardCommon.buildTotalsRow(section.totals || {}));
		card.appendChild(header);

		const tableContainer = document.createElement("div");
		card.appendChild(tableContainer);
		DashboardCommon.renderGroupTable(tableContainer, section.balances || []);

		container.appendChild(card);
	});
}

